lib/rtti.h

```diff
@@ -0,0 +1,297 @@
+            Parents::TypeInfo::dyn_cast(typeId, ptr)...};
+
+        auto it =
+            std::find_if(ptrs.begin(), ptrs.end(), [](const void *ptr) { return ptr != nullptr; });
+        return (it != ptrs.end()) ? *it : nullptr;
+    }
+};
+
+struct Base {
+    virtual ~Base() = default;
+
+    /// Queries information about dynamic type of an object.
+    /// @return typeid of current object.
+    [[nodiscard]] virtual TypeId typeId() const noexcept = 0;
+    /// Checks if given object is of specified dynamic type.
+    /// @param typeid to check with
+    [[nodiscard]] virtual bool isA(TypeId typeId) const noexcept = 0;
+
+    /// Checks if given object is of specified dynamic type.
+    /// Same as `isA`, but typeid to check with is derived from template argument.
+    template <typename T>
+    [[nodiscard]] bool is() const noexcept {
+        return isA(TypeInfo<T>::id());
+    }
+
+    /// Casts current object to a given type `T`.
+    /// Overall the semantics is similar to that of `dynamic_cast`, however,
+    /// relies on custom RTTI metadata and intrusive annotation.
+    /// @return nullptr if object is not compatible with `T` (e.g. if `T` is not
+    /// among its base classes), casted pointer to object of T* type otherwise.
+    template <typename T>
+    [[nodiscard]] T *to() noexcept {
+        return reinterpret_cast<T *>(const_cast<void *>(toImpl(TypeInfo<T>::id())));
+    }
+
+    /// Same as `to`, but returns const pointer to T.
+    template <typename T>
+    [[nodiscard]] const T *to() const noexcept {
+        return reinterpret_cast<const T *>(toImpl(TypeInfo<T>::id()));
+    }
+
+ protected:
+    [[nodiscard]] virtual const void *toImpl(TypeId typeId) const noexcept = 0;
+};
+
+}  // namespace RTTI
+
+/// Declare typeinfo for a given class `T`.
+///
+/// All direct bases (including virtual) should be specified as additional
+/// arguments. For example:
+/// struct Base1 {
+///   ...
+///   DECLARE_TYPEINFO(Base1);
+/// };
+/// struct Base2 {
+///   ...
+///   DECLARE_TYPEINFO(Base2);
+/// };
+///
+/// struct Derived : public Base1, public virtual Base2 {
+///   ...
+///   DECLARE_TYPEINFO(Derived, Base1, Base2);
+/// };
+#define DECLARE_TYPEINFO(T, ...)                                                  \
+ private:                                                                         \
+    static constexpr RTTI::TypeId static_typeId() { return RTTI::InvalidTypeId; } \
+    DECLARE_TYPEINFO_COMMON(T, ##__VA_ARGS__)
+// static_typeId is private above in order to hide explicit typeId from the base
+// class, as in e.g.
+// class A {
+//  DECLARE_TYPEINFO_WITH_TYPEID(A, 42);
+// };
+// class B : public A {
+//   DECLARE_TYPEINFO(A, B);
+// };
+// We need to ensure static_typeId is not visible inside `B`, otherwise
+// TypeIdResolver will be confused.
+
+/// Same as DECLARE_TYPEINFO but allows for explicit typeid to be specified.
+/// No further checks for typeid collision, etc. are performed. Example:
+/// struct Base {
```

Comment:
There either has to be a check that Id is not equal to `InvalidTypeId`, or (and I would prefer that just for ease of use), `static_typeId` should never be used directly in a place that expect type id.

Reply:
I think the best way to proceed is just to allow combining of hash-based Id's here? What do you think?